rclcpp/test/rclcpp/test_rosout_qos.cpp

```diff
@@ -0,0 +1,186 @@
+         lhs.lifespan == rhs.lifespan &&
+         lhs.liveliness == rhs.liveliness &&
+         lhs.liveliness_lease_duration == rhs.liveliness_lease_duration &&
+         lhs.avoid_ros_namespace_conventions == rhs.avoid_ros_namespace_conventions;
+}
+
+bool operator!=(
+  const rmw_qos_profile_t & lhs,
+  const rmw_qos_profile_t & rhs)
+{
+  return !(lhs == rhs);
+}
+
+class TestRosoutQoS : public ::testing::Test
+{
+protected:
+  void SetUp() override
+  {
+    node_options = std::make_shared<rclcpp::NodeOptions>();
+  }
+
+  void InitializeNodeOptions(const rmw_qos_profile_t rmw_qos_options = rmw_qos_profile_default)
+  {
+    rclcpp::QoS qos_profiles(rmw_qos_options);
+    node_options->rosout_qos(qos_profiles);
+  }
+
+  void TearDown()
+  {
+    node_options.reset();
+  }
+
+  std::shared_ptr<rclcpp::NodeOptions> node_options;
+};
+
+/*
+   Test rosout_qos function with default value.
+ */
+TEST_F(TestRosoutQoS, test_rosout_qos_with_default_value) {
+  rclcpp::QoS rosout_qos_profile = rclcpp::NodeOptions().rosout_qos();
+  rmw_qos_profile_t rmw_qos_profile = rosout_qos_profile.get_rmw_qos_profile();
+  EXPECT_EQ(rcl_qos_profile_rosout_default, rmw_qos_profile);
+}
+
+/*
+   Test `rosout_qos` function with user-defined value.
+ */
+TEST_F(TestRosoutQoS, test_rosout_qos_with_custom_value) {
+  rmw_time_t life_span;
+  life_span.sec = 10;
+  life_span.nsec = 0;
+  auto qos_profile = rclcpp::QoS(rclcpp::KeepLast(1000)).transient_local().lifespan(life_span);
+  auto options = rclcpp::NodeOptions().rosout_qos(qos_profile);
+  rclcpp::QoS rosout_qos = options.rosout_qos();
+  rmw_qos_profile_t rmw_qos_profile = rosout_qos.get_rmw_qos_profile();
+
+  EXPECT_EQ((size_t)1000, rmw_qos_profile.depth);
+  EXPECT_EQ(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL, rmw_qos_profile.durability);
+  EXPECT_EQ(life_span, rmw_qos_profile.lifespan);
+}
+
+/*
+   Test the QoS's constructor which has `rmw_qos_profile_t` as the parameter.
+ */
+TEST_F(TestRosoutQoS, test_qos_constructor) {
+  rmw_qos_profile_t rosout_qos_profile = rcl_qos_profile_rosout_default;
+  rclcpp::QoS qos = rclcpp::QoS(rosout_qos_profile);
+  rmw_qos_profile_t rmw_qos_profile = qos.get_rmw_qos_profile();
+  EXPECT_EQ(rcl_qos_profile_rosout_default, rmw_qos_profile);
+}
+
+/*
+   Test `get_rcl_node_options` function.
+ */
+TEST_F(TestRosoutQoS, test_get_rcl_node_options) {
+  rmw_qos_profile_t rosout_qos_profile = rcl_qos_profile_rosout_default;
+  {
+    InitializeNodeOptions(rosout_qos_profile);
+    EXPECT_EQ(rcl_qos_profile_rosout_default, node_options->get_rcl_node_options()->rosout_qos);
+  }
+
+  {
+    rosout_qos_profile.depth = 10;
+    rosout_qos_profile.durability = RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT;
+    InitializeNodeOptions(rosout_qos_profile);
+    EXPECT_NE(rcl_qos_profile_rosout_default, node_options->get_rcl_node_options()->rosout_qos);
+  }
+}
+
+/*
+   Test `rcl_node_init` function
+ */
+TEST_F(TestRosoutQoS, test_rcl_node_init) {
```

Comment:
This is supposed to be test for `rcl` not `rclcpp`? Why checking `rcl` interfaces with `rclcpp` test?

Reply:
@fujitatomoya sorry for late reply.
You are right, these test should exist in `test/rcl/test_node.cpp`, i will delete these from this file.

Reply:
This is still pending.
Opening a PR in `rcl` sounds like a great idea!